ops/main.py

```diff
@@ -42,6 +43,38 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # only allow accesing these vars because these are the currently used vars in main.py
+
+    juju_relation: str = ''
+    juju_relation_id: str = ''
+    juju_secret_revision: str = ''
+    juju_workload_name: str = ''
+    juju_action_name: Optional[str] = None
+    juju_action_uuid: Optional[str] = None
+    juju_charm_dir: Optional[str] = None
+    juju_debug: Optional[str] = None
+    juju_departing_unit: Optional[str] = None
+    juju_dispatch_path: Optional[str] = None
+    juju_notice_id: Optional[str] = None
+    juju_notice_key: Optional[str] = None
+    juju_notice_type: Optional[str] = None
+    juju_pebble_check_name: Optional[str] = None
+    juju_remote_app: Optional[str] = None
+    juju_remote_unit: Optional[str] = None
+    juju_secret_id: Optional[str] = None
+    juju_secret_label: Optional[str] = None
+    juju_storage_id: Optional[str] = ''
+
+    @classmethod
+    def from_environ(cls) -> '_JujuContext':
+        kwargs: Dict[str, Any] = {name: os.getenv(name.upper()) for name in cls.__annotations__}
```

Comment:
This is clever, but we'll probably want to go with just custom code like you've done in #1314 as we may want to customise some of the names and types.